src/inet/UDPEndPoint.cpp

```diff
@@ -543,6 +543,27 @@ INET_ERROR UDPEndPoint::SendMsg(const IPPacketInfo * pktInfo, System::PacketBuff
 
 #if CHIP_SYSTEM_CONFIG_USE_LWIP
 
+    pbuf * msgCopy = NULL;
+
+    if (sendFlags & IPEndPointBasis::kSendFlag_RetainBuffer)
```

Comment:
Could we just check whether we "own" the buffer, and do this "prepend a space for headers" bit if we don't?  Seems like checking the refcount (the same way that the assert that ends up failing does) might do the trick, right?

And then we don't need to pipe this flag through various places, sort out whether it's appropriate for its counterparent to live in PacketHeader, etc.

Does that sound feasible?  Reasonable?

Reply:
I think this is a good idea, Thanks! 

The spec says the CRMP is only used for UDP packets, I am not aware of any case that a TCP packet need to be retained if it does not use CRMP.

Reply:
> The spec says the CRMP is only used for UDP packets

I'm more thinking what happens if some consumer in general retains a buffer and then tries to send it over TCP.  Any idea what happens if they do?